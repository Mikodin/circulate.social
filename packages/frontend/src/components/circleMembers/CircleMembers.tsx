import { useState } from 'react';
import { Modal, List } from 'antd';
import { UserOutlined } from '@ant-design/icons';

export interface Props {
  circleMemberNames: string[];
  renderComponent?: JSX.Element;
}

const CircleMembers = ({
  circleMemberNames,
  renderComponent,
}: Props): JSX.Element => {
  const [isMembersModalOpen, setIsMembersModalOpen] = useState(false);

  return (
    <>
      <div onClick={() => setIsMembersModalOpen(true)}>
        {renderComponent || (
          <h4 role="button">
            <a>Members: {circleMemberNames.length}</a>
          </h4>
        )}
      </div>
      <Modal
        width={300}
        title="Members in Circle"
        visible={isMembersModalOpen}
        onOk={() => setIsMembersModalOpen(false)}
        onCancel={() => setIsMembersModalOpen(false)}
      >
        <List
          dataSource={circleMemberNames}
          renderItem={(memberName) => (
            <List.Item>
              <UserOutlined /> {memberName}
            </List.Item>
          )}
        />
      </Modal>
    </>
  );
};

export default CircleMembers;
