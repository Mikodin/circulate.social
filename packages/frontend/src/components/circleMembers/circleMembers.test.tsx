import { render, fireEvent } from '@testing-library/react';
import CircleMembers, { Props } from './CircleMembers';

const mockCircleMemberNames = ['Mike A', 'Joe T', 'Bill N'];
const DEFAULT_PROPS = {
  circleMemberNames: mockCircleMemberNames,
};

function renderContainer(overrideProps?: Partial<Props>) {
  return render(<CircleMembers {...DEFAULT_PROPS} {...overrideProps} />);
}

describe('CircleMembers', () => {
  it('Should render the h4 by default', () => {
    const container = renderContainer();
    const { queryByRole } = container;

    expect(container).toMatchSnapshot();
    expect(queryByRole('button')).toBeTruthy();
  });

  describe('When given a renderComponent', () => {
    it('Should render that component', () => {
      const container = renderContainer({
        renderComponent: <p role="custom-button">Members</p>,
      });
      const { queryByRole } = container;

      expect(container).toMatchSnapshot();
      expect(queryByRole('custom-button')).toBeTruthy();
      expect(queryByRole('button')).not.toBeTruthy();
    });
  });

  describe('When clicked', () => {
    it('Should open a Modal with the title "Members in Circle"', () => {
      const { queryByRole, queryByText } = renderContainer();
      fireEvent.click(queryByRole('button'));
      expect(queryByText('Members in Circle')).toBeTruthy();
    });
  });

  describe('The modal', () => {
    it('Should contain a list of all the members in the circle', () => {
      const { queryByRole, queryByText } = renderContainer();
      fireEvent.click(queryByRole('button'));
      mockCircleMemberNames.forEach((mockMemberName) => {
        expect(queryByText(mockMemberName)).toBeTruthy();
      });
    });
  });
});
