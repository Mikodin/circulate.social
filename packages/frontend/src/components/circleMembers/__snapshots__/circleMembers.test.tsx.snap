// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircleMembers Should render the h4 by default 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h4
          role="button"
        >
          <a>
            Members: 
            3
          </a>
        </h4>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h4
        role="button"
      >
        <a>
          Members: 
          3
        </a>
      </h4>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CircleMembers When given a render component Should render that component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p
          role="custom-button"
        >
          Members
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p
        role="custom-button"
      >
        Members
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CircleMembers When given a renderComponent Should render that component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p
          role="custom-button"
        >
          Members
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p
        role="custom-button"
      >
        Members
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
